tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -270,109 +270,230 @@ void EnqueueProgramCommand::assemble_preamble_commands() {
-            uint32_t runtime_args_len = transfer_info[packed_cmd_idx].data.size();
-
-            uint32_t num_packed_cmds_in_seq = 0;
-
-            for (int i = packed_cmd_idx; i < num_packed_cmds; i++) {
-                TT_ASSERT(transfer_info[i].dst_noc_info.size() == 1, "Not supporting CoreRangeSet for runtime args");
-                uint32_t max_runtime_args_len = std::max(runtime_args_len, (uint32_t)transfer_info[i].data.size());
-                uint32_t next_cmd_sequence_sizeB = align(
-                    sizeof(CQPrefetchCmd) + get_runtime_payload_sizeB(num_packed_cmds_in_seq + 1, max_runtime_args_len, true), PCIE_ALIGNMENT);
-
+    constexpr bool unicast = std::is_same<PackedSubCmd, CQDispatchWritePackedUnicastSubCmd>::value;
+
+    uint32_t num_packed_cmds_in_seq = sub_cmds[idx].size();
+    uint32_t max_packed_cmds = get_max_num_packed_cmds(max_runtime_args_len[idx], max_prefetch_command_size, unicast);
+    uint32_t offset_idx = 0;
+    program.command_indices.processor_to_cmd_mapping[idx] = {};
+    while (num_packed_cmds_in_seq != 0) {
+        uint32_t num_packed_cmds = std::min(num_packed_cmds_in_seq, max_packed_cmds);
+        uint32_t rt_payload_sizeB =
+            get_runtime_payload_sizeB(num_packed_cmds, max_runtime_args_len[idx], unicast);
+        uint32_t cmd_sequence_sizeB = align(sizeof(CQPrefetchCmd) + rt_payload_sizeB, PCIE_ALIGNMENT);
+        DeviceCommand command_sequence(cmd_sequence_sizeB);
+        command_sequence.add_dispatch_write_packed<PackedSubCmd>(
+            num_packed_cmds,
+            l1_arg_base_addr[idx],
+            max_runtime_args_len[idx] * sizeof(uint32_t),
+            rt_payload_sizeB,
+            sub_cmds[idx],
+            rt_data_and_sizes[idx],
+            offset_idx);
+        runtime_args_command_sequences.emplace_back(command_sequence);
+
+        program.command_indices.processor_to_cmd_mapping[idx].emplace_back(
+            std::array<uint32_t, 4>{(uint32_t)runtime_args_command_sequences.size() - 1,
+            num_packed_cmds,
+            (uint32_t)get_runtime_args_data_offset(num_packed_cmds, max_runtime_args_len[idx], unicast),
+            align(max_runtime_args_len[idx] * sizeof(uint32_t), L1_ALIGNMENT)
+        });
+        program.cached_runtime_args_commands.emplace_back(command_sequence.cmd_vector());
+        num_packed_cmds_in_seq -= num_packed_cmds;
+        offset_idx += num_packed_cmds;
+    }
+}
 
-                if (next_cmd_sequence_sizeB > max_prefetch_command_size) {
-                    // If we hit this, then we send multiple write packed commands for each RT dst.
-                    // TODO: Revisit RT arg command sequence construction to be more performant
-                    break;
+// Generate command sequence for unique (unicast) and common (multicast) runtime args
+void EnqueueProgramCommand::assemble_runtime_args_commands() {
+    static const std::vector<uint32_t> processor_to_l1_arg_base_addr = {
```

Comment:
can you leave a comment here that links to this:
```
enum class RISCV : uint8_t {
    BRISC = 0,
    NCRISC = 1,
    TRISC0 = 2,
    TRISC1 = 3,
    TRISC2 = 4,
    ERISC = 5,
    COMPUTE = 6,     // Encompasses TRISC0, TRISC1, and TRISC2
};
```

Reply:
Done